contracts/VaultV3.vy

```diff
@@ -141,75 +225,131 @@ def erc20_safe_transfer(token: address, receiver: address, amount: uint256):
     self.totalIdle += amount
 
-    log Deposit(_recipient, shares, amount)
+    log Deposit(_fnCaller, _recipient, amount, shares)
 
     return shares
 
-@external
-def redeem(_shares: uint256 = MAX_UINT256, _recipient: address = msg.sender, _strategies: DynArray[address, 10] = []) -> uint256:
-    owner: address = msg.sender
-    shares: uint256 = _shares
-
-    if _shares == MAX_UINT256:
-        sharesBalance: uint256 = self.balanceOf[owner]
-        shares = sharesBalance
-
-    assert shares > 0, "no shares to withdraw"
+@internal
+def _redeem(fnCaller: address, _receiver: address, _owner: address, _shares: uint256, _strategies: DynArray[address, 10] = []) -> uint256:
+    if fnCaller != _owner: 
+      self._spendAllowance(_owner, msg.sender, _shares)
 
-    amount: uint256 = self._amountForShares(shares)
+    assets: uint256 = self._amountForShares(_shares)
 
     # TODO: withdraw from strategies
     
     # TODO: gas savings: totalIdle should be cached if used above
-    assert self.totalIdle >= amount, "insufficient total idle"
+    assert self.totalIdle >= assets, "insufficient total idle"
 
-    self._burnShares(shares, owner)
-    self.totalIdle -= amount
+    self._burnShares(_shares, _owner)
 
-    self.erc20_safe_transfer(ASSET.address, _recipient, amount)
+    self.totalIdle -= assets
+    self.erc20_safe_transfer(ASSET.address, _receiver, assets)
 
-    log Withdraw(_recipient, shares, amount)
+    log Withdraw(msg.sender, _receiver, _owner, _shares, assets)
 
-    return amount
+    return assets
+
+
+## ERC4626 ## 
+@external
+def asset() -> address:
+   return ASSET.address
 
-# SHARE MANAGEMENT FUNCTIONS #
 @view
 @external
 def totalAssets() -> uint256:
     return self._totalAssets()
 
-@view
 @external
-def pricePerShare() -> uint256:
-    return self._amountForShares(10 ** self.decimals)
+def convertToShares(assets: uint256) -> uint256:
+   return self._sharesForAmount(assets)
+
+@external
+def convertToAssets(shares: uint256) -> uint256:
+   return self._amountForShares(shares)
+
+@external
+def maxDeposit(owner: address) -> uint256:
+   return MAX_UINT256
+
+@external
+def maxMint(owner: address) -> uint256: 
+   maxDeposit: uint256 = MAX_UINT256
+   return self._sharesForAmount(maxDeposit)
+
+@external
+def maxWithdraw(owner: address) -> uint256:
+   # TODO: calculate max between liquidity and shares
+   return 0
+
+@external
+def maxRedeem(owner: address) -> uint256:
+   # TODO: add max liquidity calculation
+   return self.balanceOf[owner]
 
 @external
-def sharesForAmount(amount: uint256) -> uint256:
-    return self._sharesForAmount(amount)
+def previewDeposit(assets: uint256) -> uint256:
+   return self._sharesForAmount(assets)
 
 @external
-def amountForShares(shares: uint256) -> uint256:
+def previewWithdraw(shares: uint256) -> uint256:
     return self._amountForShares(shares)
 
+@external
+def deposit(assets: uint256, receiver: address) -> uint256:
```

Comment:
doesn't implementing 4626 in here mean that we now only have one deposit interface? this means we can't have separate deposit functions for different integrators, and that they will all have to integrate through 4626 (and will just have to use a default withdrawal queue). is this desirable?

Reply:
it doesn't mean that imo. We can overload deposit so it's compliant with ERC4626 and have other more flexible ways of withdrawing. Wdyt?

Reply:
i don't think vyper has the same overloading features that solidity has: 

https://github.com/vyperlang/vyper/issues/903

vyper has a more limited overloading where you can set default values for variables. in this case we would have to set a default value for the queue of strategies to withdraw.

i think the main challenge of having ERC4626 joined with the vault code is that if we have a periphery contract that returns a default queue, then the vault will now depend on that contract. if our goal is to to have the vault be immutable with as few external dependencies as possible, then this is probably not the right way to go, unless you want to manage the default queue within the vault as well. in which case we can just use the overloading described above with a vault-managed default strategy queue.

Reply:
I agree having a separated withdrawal queue may not be the best approach. 

I think we should overload withdraw function and have a version that allows the sender to specify which strategies to withdraw from. 
`def withdraw(_assets: uint256, _receiver: address, _owner: address, _strategiesToWithdraw: DynArray[address, 10] = []) -> uint256`
so that the custom and the ERC4626 compliant withdraw (and redeem) functions are available. 

For withdrawal queue, we have 3 options for simple ERC4626 compliance: 
- assume ERC4626 can only withdraw from funds in the vault (this is simpler, lower gas and 
- implement a simple withdrawal queue with low gas, high liquidity set of strategies to withdraw from
- implement a simple withdrawal queue with all strategies like in V2

Wdyt? @pandadefi @anticlimactic

Reply:
> implement a simple withdrawal queue with low gas, high liquidity set of strategies to withdraw from

how would this work? who would choose the strategies that we are withdrawing from? (i.e. will it be done in a centralized way or decentralized way?) i'm mindful of the direction we are heading in so want to make sure we're being consistent on our path to being more decentralized.

i don't think option 1 is a good idea. option 2 or option 3 are better. and the reason why is because option 1 would make using the vault via erc4626 too restrictive (i.e. it would not be feature-complete, if you compare it to using the vault through a custom interface). if we go with opt 1 (and to some degree opt 2), the vault becomes less composable. consider generalized lending protocols that may want to deposit and withdraw from our vault -- if not all strategies are included, it may be impossible for them withdraw without some intervention. more likely, they would have to build a custom adapter, which would defeat the purpose of implementing erc4626 in the first place.

Reply:
i do not believe our opinions conflict -- i just think there should be a way for people to withdraw from the vault, through erc4626, without having to depend on strategists freeing funds (assuming funds are withdrawable). this would make option 2 the most workable solution. option 1 doesn't allow for this (as having totalIdle always available means we are not being capital efficient). and option 3 is gas intensive.